[jit] Fix value association with dictionaries in the tracer

suo opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* **#40885 [jit] Fix value association with dictionaries in the tracer**

`TracingState::setValue` associates a concrete IValue in the traced
program with a `Value*` symbolic. Previously, the logic for how
GenericDicts worked was special cased to only work for very simple cases
and silently eat other cases.

This PR generalizes the logic to reflect the same behavior as using
dictionaries on input: whenever we encounter a dictionary in the system,
we completely "burn in" all the keys into the graph, and then
recursively call `setValue` on the associated value.

This has the effect of requiring that any dictionary structure you are
creating in a traced program be of fixed structure, similar to how any
dictionary used as input must be static as well.

Differential Revision: [D22342490](https://our.internmc.facebook.com/intern/diff/D22342490)

suo committed:
[jit] Fix value association with dictionaries in the tracer

`TracingState::setValue` associates a concrete IValue in the traced
program with a `Value*` symbolic. Previously, the logic for how
GenericDicts worked was special cased to only work for very simple cases
and silently eat other cases.

This PR generalizes the logic to reflect the same behavior as using
dictionaries on input: whenever we encounter a dictionary in the system,
we completely "burn in" all the keys into the graph, and then
recursively call `setValue` on the associated value.

This has the effect of requiring that any dictionary structure you are
creating in a traced program be of fixed structure, similar to how any
dictionary used as input must be static as well.

[ghstack-poisoned]

wanchaol commented on torch/csrc/jit/frontend/tracer.cpp:
```diff
@@ -509,28 +509,11 @@ void TracingState::setValue(const IValue& v, Value* value) {
-    }
-    if (value_type->cast<TensorType>()) {
-      handle_unpack(aten::values);
+    for (const auto& entry : dict) {
```
This will actually implies that if dict[tensor, tensor] (tensor as keys) as input or cross tracing/scripting, the tensor key will always be treated as constant, which is ok, but I am wondering if we should throw some warning here to user?

suo replied:
Agreed, I filed https://github.com/pytorch/pytorch/issues/40886 as a followup to improve the error messaging in all cases where a key mismatch occurs

wanchaol commented on torch/csrc/jit/frontend/tracer.cpp:
```diff
@@ -509,28 +509,11 @@ void TracingState::setValue(const IValue& v, Value* value) {
+      setValue(entry.value(), static_value);
     }
   } else {
     std::ostringstream os;
```
error message seems out dated

wanchaol reviewed (approved):
looks good!

suo committed:
Update on "[jit] Fix value association with dictionaries in the tracer"


`TracingState::setValue` associates a concrete IValue in the traced
program with a `Value*` symbolic. Previously, the logic for how
GenericDicts worked was special cased to only work for very simple cases
and silently eat other cases.

This PR generalizes the logic to reflect the same behavior as using
dictionaries on input: whenever we encounter a dictionary in the system,
we completely "burn in" all the keys into the graph, and then
recursively call `setValue` on the associated value.

This has the effect of requiring that any dictionary structure you are
creating in a traced program be of fixed structure, similar to how any
dictionary used as input must be static as well.

Differential Revision: [D22342490](https://our.internmc.facebook.com/intern/diff/D22342490)

[ghstack-poisoned]
